src/compiler/scala/tools/nsc/backend/jvm/ClassHandler.scala

```diff
@@ -0,0 +1,298 @@
+            case Some(r:Runnable) => r.run()
+            case None => Await.ready(fut, Duration.Inf)
+        }
+        //we know that they are complete by we need to check for exception
+        //but first get any reports
+        unitResult.relayReports(directBackendReporting)
+        fut.value.get.get
+      }
+
+      // This way it is easier to test, as the results are deterministic
+      // the the loss of potential performance is probably minimal
+      pending().foreach {
+        unitResult: UnitResult =>
+          try {
+            stealWhileWaiting(unitResult, unitResult.task)
+            stealWhileWaiting(unitResult, unitResult.result.future)
+          } catch {
+            case NonFatal(t) =>
+              t.printStackTrace()
+              postProcessor.bTypes.frontendAccess.backendReporting.error(NoPosition, s"unable to write ${unitResult.source} $t")
+          }
+      }
+    }
+    def tryStealing:Option[Runnable]
+  }
+  private final class SyncWritingClassHandler(val unitInfoLookup: UnitInfoLookup,
+                val postProcessor: PostProcessor, val cfWriter: ClassfileWriter)
+    extends WritingClassHandler((r) => r.run()) {
+
+    override def toString: String = s"SyncWriting [$cfWriter]"
+
+    override def tryStealing: Option[Runnable] = None
+  }
+  private final case class ExecutorServiceInfo( maxThreads:Int, javaExecutor : ExecutorService, queue:BlockingQueue[Runnable])
+
+  private final class AsyncWritingClassHandler(val unitInfoLookup: UnitInfoLookup,
+                                               val postProcessor: PostProcessor,
+                                               val cfWriter: ClassfileWriter,
+                                               val executorServiceInfo: ExecutorServiceInfo)
+    extends WritingClassHandler(executorServiceInfo.javaExecutor) {
+
+    override def toString: String = s"AsyncWriting[additional threads:${executorServiceInfo.maxThreads} writer:$cfWriter]"
+
+    override def close(): Unit = {
+      super.close()
+      executorServiceInfo.javaExecutor.shutdownNow()
+    }
+
+    override def tryStealing: Option[Runnable] = Option(executorServiceInfo.queue.poll())
+  }
+
+}
+//we avoid the lock on frontendSync for the common case, when compiling to a single target
+sealed trait UnitInfoLookup {
+  def outputDir(source:AbstractFile) : AbstractFile
+  val frontendAccess: PostProcessorFrontendAccess
+}
+final class SingleUnitInfo(val frontendAccess: PostProcessorFrontendAccess, constantOutputDir:AbstractFile) extends UnitInfoLookup {
+  override def outputDir(source: AbstractFile) = constantOutputDir
+}
+final class LookupUnitInfo(val frontendAccess: PostProcessorFrontendAccess) extends UnitInfoLookup {
+  lazy val outputDirectories = frontendAccess.compilerSettings.outputDirectories
+  override def outputDir(source: AbstractFile) = outputDirectories.outputDirFor(source)
+}
+sealed trait SourceUnit {
+  def withBufferedReporter[T](fn: => T): T
```

Comment:
This method is not used, so the `localReporter` is never set

Reply:
looks to be a missing commit. Will investigate. Should be set when we execute in background

Also needs a unit test